core/src/main/java/hudson/model/AbstractBuild.java

```diff
@@ -880,15 +880,12 @@ public EnvVars getEnvironment(TaskListener log) throws IOException, InterruptedE
         if (ws!=null)   // if this is done very early on in the build, workspace may not be decided yet. see HUDSON-3997
             env.put("WORKSPACE", ws.getRemote());
 
-        project.getScm().buildEnvVars(this,env);
+        project.getScm().buildEnvironment(this, env);
 
         if (buildEnvironments!=null)
             for (Environment e : buildEnvironments)
                 e.buildEnvVars(env);
 
-        for (EnvironmentContributingAction a : getActions(EnvironmentContributingAction.class))
-            a.buildEnvVars(this,env,getBuiltOn());
```

Comment:
The `resolve` call on the next line is supposed to handle recursive definitions. Thus can you can

```java
envs.put("MYDIR", "${WORKSPACE}/subdir");
```

> My proposal would be to not touch this invocation if possible

By leaving both calls in place, we *are* touching the invocation: in 2.75 we were contributing variables once, in 2.76 twice (for freestyle builds). That seems much more likely to cause regressions (and will certainly hurt performance a little bit).